<a href="https://colab.research.google.com/github/kim0129s/py08/blob/master/notebook/ch8_exercises_py08.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ch8_exercises_py08.ipynb

Q1. 문자열 바꾸기
> 다음과 같은 문자열이 있다.

```
a:b:c:d
```

> 문자열의 split와 join 함수를 사용하여 위 문자열을 다음과 같이 고치시오.

```
a#b#c#d
```


In [3]:
# [Q1] Code-1 
a = "a:b:c:d"
b = a.split(":") 
c = "#".join(b)

print(b)
c 

['a', 'b', 'c', 'd']


'a#b#c#d'

In [4]:
# [Q1] Code-2
a = 'a:b:c:d'
colon= ':'
st_a = a.split(colon)
sharp = '#'

print(st_a)
sharp.join(st_a)

['a', 'b', 'c', 'd']


'a#b#c#d'

## Q2. 딕셔너리 값 추출하기
> 다음은 딕셔너리의 a에서 'C'라는 key에 해당하는 value를 출력하는 프로그램이다.
```
>>> a = {'A':90, 'B':80}
>>> a['C']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'C'
```
>> a 딕셔너리에는 'C'라는 key가 없으므로 위와 같은 오류가 발생한다.  
>>'C'에 해당하는 key 값이 없을 경우 오류 대신 70을 얻을 수 있도록 수정하시오.

In [ ]:
# [Q2] Code-1
a = {'A':90, 'B':80}
a['C'] = '70' 
print(a) 
a['C']


{'A': 90, 'B': 80, 'C': '70'}


'70'

In [ ]:
# [Q2] Code-2
a = {'A':90, 'B':80}
a.get('C', 70)  


70

## Q3. 리스트의 더하기와 extend 함수
> 다음과 같은 리스트 a가 있다.
```
a = [1, 2, 3]
```
> 리스트 a에 [4, 5]를 + 기호를 사용하여 더한 결과는 다음과 같다.
```
>>> a = [1, 2, 3]
>>> a = a + [4,5]
>>> a
[1, 2, 3, 4, 5]
```
> 리스트 a에 [4,5]를 extend를 사용하여 더한 결과는 다음과 같다.
```
>>> a = [1, 2, 3]
>>> a.extend([4, 5])
>>> a
[1, 2, 3, 4, 5]
```

> ### + 기호를 사용하여 더한 것과 extend한 것의 차이점이 있을까? 있다면 그 차이점을 설명하시오.

#### [Q3] Explain (코드와 주석으로 설명하시오)


### Code-1
- 리스트 a에 + 기호를 사용하는 경우

In [ ]:
# Code-1
a = [1, 2, 3]
id(a)

140599382363272

In [ ]:
a = a + [4,5]
a

[1, 2, 3, 4, 5]

In [ ]:
id(a)

140599381863624

### Code-1 설명
> ```
a = [1, 2, 3]
id(a)
=> 140599382363272
``` 
> - id 함수는 입력 받은 리스트 주소 값을 돌려준다. 
- 현재 리스트 a는 주소 140599382363272에 저장되어 있다.

> ```
a = a + [4,5]
a
=> [1, 2, 3, 4, 5]
```
> - 리스트 a에 + 기호를 사용하여 리스트 값을 추가한다. 

>```
id(a)
=> 140599381863624
```
> - 결과: 리스트 값을 추가하기 전과 다른 주소 값을 돌려준다. 주소 값이 다르므로 리스트 a는 값이 변하는 것이 아니라 새로운 리스트가 반환된다는 것임을 알 수 있다.







### Code-2
- 리스트 a에 extend 함수를 사용한 경우

In [ ]:
# Code-2
a = [1,2,3]
id(a)

139690869183048

In [ ]:
a.extend([4,5])
a

[1, 2, 3, 4, 5]

In [ ]:
id(a)

139690869183048

### Code-2 설명
> ```
a = [1, 2, 3]
id(a)
=> 139690869183048
``` 
> - 현재 리스트 a는 주소 값은 139690869183048이다.

> ```
a.extend([4,5])
a
=> [1, 2, 3, 4, 5]
```
> - 리스트 a에 extend 함수를 사용하여 [4,5]라는 값을 추가한다. 

>```
id(a)
=> 139690869183048
```
> - 결과: '+' 기호를 사용하여 더한 경우와 다르게 주소값이 변하지 않고 그대로 유지되는 것을 알 수 있다.

## Q4. 리스트 총합 구하기
> 다음은 A학급 학생의 점수를 나타내는 리스트이다. 다음 리스트에서 50점 이상 점수의 총합을 구하시오.
```
A = [20, 55, 67, 82, 45, 33, 90, 87, 100, 25]
```

In [ ]:
# [Q4] Code-1
A = [20, 55, 67, 82, 45, 33, 90, 87, 100, 25]

result = 0
for i in A:
  if i >= 50:
      result += i

print(result)


481


> - pop함수 사용

In [ ]:
# [Q4] Code-2
A = [20, 55, 67, 82, 45, 33, 90, 87, 100, 25]

result = 0
while A:                
    i = A.pop()      
    if i >= 50:      
        result += i

print(result)


481


In [ ]:
# [Q4] Code-3
A = [20, 55, 67, 82, 45, 33, 90, 87, 100, 25]
num = -1
sum = 0
while num < 9:
	num += 1
	if A[num] < 50: continue
	sum = sum + A[num]
print(sum)

481


## Q5. 피보나치 함수
> 첫 번째 항의 값이 0이고 두 번째 항의 값이 1일 때, 이후에 이어지는 항은 이전의 두 항을 더한 값으로 이루어지는 수열을 피보나치 수열이라고 한다.
```
0, 1, 1, 2, 3, 5, 8, 13, 21, ...
```
> 입력을 정수 n으로 받았을 때, n 이하까지의 피보나치 수열을 출력하는 함수를 작성해 보자.


In [ ]:
# [Q5] Code-1
def fibo(n):
    return fibo(n-1) + fibo(n-2) if n>=2 else n 
    # n이 0이나 1일 때는 값도 0,1이기 때문에 그대로 반환 

for n in range(1,11):
    print(n, fibo(n))  # n번째 , n번째 피보나피 수 

1 1
2 1
3 2
4 3
5 5
6 8
7 13
8 21
9 34
10 55


In [ ]:
# [Q5] Code-2
def fibo(n):
    if n<2:  # n이 2 미만일 때는 그대로 반환
      return n
    a,b = 0,1
    for i in range(n-1):  # n이 2 이상일 때는 n-1번 만큼 반복문 실행
        a,b = b,a+b
    return b
for n in range(1,11):
    print(n,fibo(n))

1 1
2 1
3 2
4 3
5 5
6 8
7 13
8 21
9 34
10 55


In [ ]:
# [Q5] Code-3
def fib(n):
    if n == 0 : return 0         # n이 0,1 일때는 값도 0,1
    if n == 1 : return 1          
    return fib(n-2) + fib(n-1)   # n이 2 이상일 때는 그 이전의 두 값을 더하여 반환

for i in range(11):
    print(fib(i))


0
1
1
2
3
5
8
13
21
34
55


In [ ]:
# [Q5] Code-4
def fibo(n):
   
    if n<1 : return 0
    if n==1 or n==2 : return 1
   
   
    return fibo(n-1)+fibo(n-2)
 
N=int(input('수를 입력하세요 : '))
 
for i in range(1,N+1):
    print(fibo(i))

수를 입력하세요 : 10
1
1
2
3
5
8
13
21
34
55


## Q6. 숫자의 총합 구하기
> 사용자로부터 다음과 같은 숫자를 입력받아 입력받은 숫자의 총합을 구하는 프로그램을 작성하시오. (단 숫자는 콤마로 구분하여 입력한다.)

```
65,45,2,3,45,8
```


In [ ]:
# [Q6] Code-1
number_input = input("숫자를 입력하세요: ")
numbers = number_input.split(",")

def sum_number(numbers,total=0):
    for num in numbers:
      total += int(num)
    return total

print(sum_number(numbers))

숫자를 입력하세요: 65, 45, 2, 3, 45, 8
168


In [ ]:
# [Q6] Code-2
number_input = input("숫자를 입력하세요: ")
numbers = number_input.split(",")
total = 0
for num in numbers:
    total += int(num)    # 입력은 문자열이므로 숫자로 변환해야 한다.
print(total)


숫자를 입력하세요: 65, 45, 2, 3, 45, 8
168


> - map함수 사용

In [ ]:
# [Q6] Code-3
number_list=[]
number_list = map(int, input('숫자를 입력하세요: ').split(","))
total = 0
for num in number_list:
    total += int(num)
print(total)

숫자를 입력하세요: 65,45,2,3,45,8
168


## Q7. 한 줄 구구단
> 사용자로부터 2~9의 숫자 중 하나를 입력받아 해당 숫자의 구구단을 한 줄로 출력하는 프로그램을 작성하시오.

> 실행 예)
```
구구단을 출력할 숫자를 입력하세요(2~9): 2
2 4 6 8 10 12 14 16 18
```


In [ ]:
# [Q7] Code-1
gugudan_input = input("구구단을 출력할 숫자를 입력하세요(2~9):")
dan = int(gugudan_input) 

def gugudan(dan):
    for n in range(1,10):
        print(f'{dan * n}',end=' ')

gugudan(dan)


구구단을 출력할 숫자를 입력하세요(2~9):2
2 4 6 8 10 12 14 16 18 

In [ ]:
# [Q7] Code-2     
gugudan_input = input("구구단을 출력할 숫자를 입력하세요(2~9):")
dan = int(gugudan_input) 

for i in range(1, 10):
    print(i*dan, end= ' ')  


구구단을 출력할 숫자를 입력하세요(2~9):2
2 4 6 8 10 12 14 16 18 

In [ ]:
# [Q7] Code-3
gugudan_input = input("구구단을 출력할 숫자를 입력하세요(2~9):")

list1 = [ n*int(dan) for n in range(1,10) for dan in gugudan_input ]
print(list1)


구구단을 출력할 숫자를 입력하세요(2~9):2
[2, 4, 6, 8, 10, 12, 14, 16, 18]


In [ ]:
# [Q7] Code-4
gugudan = int(input("구구단을 출력할 숫자를 입력하세요(2~9): "))

if gugudan<=10 and gugudan>=1:
    for n in range(1,10):
      print(format(gugudan*n), end=' ')

구구단을 출력할 숫자를 입력하세요(2~9): 2
2 4 6 8 10 12 14 16 18 

## Q8. 역순 저장
> 다음과 같은 내용의 파일 abc.txt가 있다.
```
AAA
BBB
CCC
DDD
EEE
```
> 이 파일의 내용을 다음과 같이 역순으로 바꾸어 저장하시오.
```
EEE
DDD
CCC
BBB
AAA
```

In [ ]:
# [Q8] Code-1
f = open('abc.txt','r')
lines = f.readlines()
f.close()

In [ ]:
for lines in reversed(open("abc.txt").readlines()):
    print(lines.rstrip())

EEE
DDD
CCC
BBB
AAA


In [ ]:
# [Q8] Code-2
f = open('abc.txt', 'r')
lines = f.readlines()    # 모든 라인을 읽음
f.close()

lines.reverse()          # 읽은 라인을 역순으로 정렬

f = open('abc.txt', 'w')
for line in lines:
    line = line.strip()  # 포함되어 있는 줄 바꿈 문자 제거
    f.write(line)
    f.write('\n')        # 줄 바꿈 문자 삽입
f.close()

In [ ]:
f = open("/content/abc.txt",'r')
line=f.read()
print(line)
f.close()

EEE
DDD
CCC
BBB
AAA



## Q9. 평균값 구하기
> 다음과 같이 총 10줄로 이루어진 sample.txt 파일이 있다. sample.txt 파일의 숫자 값을 모두 읽어 총합과 평균 값을 구한 후 평균 값을 result.txt 파일에 쓰는 프로그램을 작성하시오.
```
70
60
55
75
95
90
80
80
85
100
```


### Code-1

In [ ]:
# [Q9] Code-1
f = open("sample.txt",'r')
score = f.readlines()
score = list(map(int, score))
score_sum=0

for i in score:
    score_sum = score_sum +  i
score_average = score_sum / len(score)

print("총합 = %d"%score_sum)
print("평균 = %.2f"%score_average)


총합 = 790
평균 = 79.00


> - result.txt를 생성하고 score_average 값 쓰기

In [ ]:
f = open('result.txt','w')
f.write(str(score_average))
f.close()

> - result.txt에 쓴 score_average 값 읽기

In [ ]:
f = open("result.txt",'r')
line=f.read()
print(line)
f.close()

79.0


### Code-2

In [ ]:
# [Q9] Code-2
f = open("sample.txt")
lines = f.readlines()
f.close()

total = 0
for line in lines:
    score = int(line)
    total += score
average = total / len(scores)

print(total)
print(average)


790
79.0


> - result.txt를 생성하고 average 값 쓰고 읽기

In [ ]:
f = open ("result.txt","w")
f.write(str(average))
f.close()

f = open("result.txt","r")
average = f.read()
print(average)
f.close()

79.0


## Q10. 사칙연산 계산기
> 다음과 같이 동작하는 클래스 Calculator를 작성하시오.
```
>>> cal1 = Calculator([1,2,3,4,5])
>>> cal1.sum() # 합계
15
>>> cal1.avg() # 평균
3.0
>>> cal2 = Calculator([6,7,8,9,10])
>>> cal2.sum() # 합계
40
>>> cal2.avg() # 평균
8.0
```


In [ ]:
# [Q10] Code-1
class Calculator:
    def __init__ (self, numberList): 
        self.numberList = numberList
        
    def add(self): 
        result = 0
        for num in self.numberList: 
            result += num
        return result

    def avg(self):
        total = self.add()
        return total / len(self.numberList)

cal1 = Calculator([1,2,3,4,5]) 
print (cal1.add())
print (cal1.avg())

cal2 = Calculator([6,7,8,9,10]) 
print (cal2.add())
print (cal2.avg())

15
3.0
40
8.0


In [ ]:
# [Q10] Code-2
class Calculator:
    def __init__(self,value):
        self.value = value
 
    def add(self):
        result = 0
        for num in self.value: 
            result += num
        return result

class UpgradeCalculator(Calculator):
    def avg(self):
        total = self.add()
        return total / len(self.value)

cal1 = UpgradeCalculator([1,2,3,4,5])
print(cal1.add())
print(cal1.avg())

cal2=UpgradeCalculator([6,7,8,9,10])
print(cal2.add())
print(cal2.avg())

15
3.0
40
8.0


## Q11. 모듈 사용 방법
> ./doit 디렉터리에 mymod.py 파이썬 모듈이 있다고 가정해 보자. 명령 프롬프트 창에서 파이썬 셸을 열어 이 모듈을 import해서 사용할 수 있는 방법을 모두 기술하시오. (즉 다음과 같이 import mymod를 수행할 때 오류가 없어야 한다.)
```
>>> import mymod
>>>
```

In [ ]:
!mkdir doit
# 실행 후 새로 고침!
# mymod.py 를 업로드한다.

In [ ]:
!mv mymod.py ./doit
# 실행 후 새로 고침!

#### 모듈 사용 방법

- sys.path.append()를 사용하여 새 경로를 추가

In [ ]:
# [Q11] Code-1 
import sys
sys.path.append("/content/doit")
import mymod

- from 모듈 import 변수나 함수

In [ ]:
# [Q11] Code-2
from doit import mymod

- 현재 디렉터리를 doit으로 변경

In [ ]:
# [Q11] Code-3
import os
os.chdir("/content/doit")  

- pwd 리눅스 명령어와 같은 결과 => '/content/doit'

In [ ]:
# [Q11] Code-4
os.getcwd()  
import mymod

## Q12. 오류와 예외 처리
> 다음 코드의 실행 결과를 예측하고 그 이유에 대해 설명하시오.

```
result = 0

try:
    [1, 2, 3][3]
    "a"+1
    4 / 0
except TypeError:
    result += 1
except ZeroDivisionError:
    result += 2
except IndexError:
    result += 3
finally:
    result += 4

print(result)
```


In [ ]:
# [Q12] Explain (코드와 주석으로 설명하시오)
result = 0
 
try:
    [1, 2, 3][3]
    "a"+1
    4 / 0
except TypeError:
    result += 1
except ZeroDivisionError:
    result += 2
except IndexError:
    result += 3
finally:
    result += 4
 
print(result)


7


### [Q12] Explain
 - result의 초깃값은 0이다.
>```
result = 0
```

- try문 안의 [1,2,3][3]이라는 문장 수행 시 IndexError가 발생하여 except IndexError: 구문으로 이동하게 되어 result에 3이 더해져 3이 된다.
>```
try:
    [1, 2, 3][3]
    "a"+1
    4 / 0
    --> IndexError 발생
``` 
>```
except IndexError:
    result += 3
```

- 최종적으로 finally 구문이 실행되어 result에 4가 더해져 7이 된다.
>```
finally:
    result += 4
```

- print(result)가 수행되어 result의 최종 값인 7이 출력이 된다.
>```
print(result)
>>>> 7
```

## Q13. DashInsert 함수
> DashInsert 함수는 숫자로 구성된 문자열을 입력받은 뒤 문자열 안에서 홀수가 연속되면 두 수 사이에 - 를 추가하고, 짝수가 연속되면 * 를 추가하는 기능을 갖고 있다. DashInsert 함수를 완성하시오.
```
입력 예시: 4546793
출력 예시: 454*67-9-3
```


In [ ]:
# [Q13] Code-1
data = "4546793"
numbers = list(map(int, data))   # 숫자 문자열을 숫자 리스트로 변경
result = []

for i, num in enumerate(numbers):
    result.append(str(num))
    if i < len(numbers)-1:                    # 다음 수가 있다면
        is_odd = num % 2 == 1                 # 현재 수가 홀수
        is_next_odd = numbers[i+1] % 2 == 1   # 다음 수가 홀수
        if is_odd and is_next_odd:            # 연속 홀수
            result.append("-")                
        elif not is_odd and not is_next_odd:  # 연속 짝수  
            result.append("*")

print("".join(result))


454*67-9-3


In [ ]:
# [Q13] Code-2
number_list = input("숫자를 입력해주세요: ")

result = number_list[0]
for a,b in zip(number_list[0:], number_list[1:]): 
    if int(a)%2 == 0 and int(b)%2 == 0:    # 연속 짝수
        result += '*' + b
    elif int(a)%2 == 1 and int(b)%2 == 1:  # 연속 홀수
        result += '-' + b
    else:
        result += b                        #홀수짝수, 짝수홀수

print(result)


숫자를 입력해주세요: 4546793
454*67-9-3


In [ ]:
# [Q13] Code-3
number_list = list(map(int,' '.join(input("숫자를 입력해주세요: ")).split()))
result = [str(number_list[0])]
for x in range(len(number_list)-1):
    if number_list[x]%2==0 and number_list[x+1]%2==0:
        result.append('*')
    if number_list[x]%2==1 and number_list[x+1]%2==1:
        result.append('-')
    result.append(str(number_list[x+1]))
print(''.join(result))

숫자를 입력해주세요: 4546793
454*67-9-3


## Q14. 문자열 압축하기
> 문자열을 입력받아 같은 문자가 연속적으로 반복되는 경우에 그 반복 횟수를 표시해 문자열을 압축하여 표시하시오.
```
입력 예시: aaabbcccccca
출력 예시: a3b2c6a1
```



In [ ]:
# [Q14] Code-1
def zip(sentence):
    count = 0  # 문자 카운팅
    next_sentence = "" # sentence 감지
    result = ""  # 결과 저장
    for i in sentence:
        if next_sentence != i:    # 문자가 달라지는 곳을 감지 ===> a다음 b의 차례가 오면 if문 실행
            next_sentence = i     # 문자열 첫 문자는 a
            if count:             # 이때 count = 0 이므로 False
                result += str(count)
            result += i           # 공백 result에 a가 추가
            count = 1             # count = 1이 매겨짐
        else:                     # 다음 문자도 a이기 때문에 else문으로 넘어가고 count = 2
          count += 1              # 이과 같은 방법으로 반복문 실행
    if count:
        result += str(count)      # b의 차례가 시작되면 a의 count가 3이므로 result에 입력 => a3
    return result

sentence = 'aaabbcccccca'
print(zip(sentence))

a3b2c6a1


In [ ]:
# [Q14] Code-2
sentence = "aaabbcccccca" # 초기값 설정
count = 1                 # 문자 카운팅
count_sentence =[]        # 결과 저장

for i in range(len(sentence)-1):
    if sentence[i] == sentence[i+1]: # 현재 문자와 다음 문자와 같으면 count에 1 더하기
        count += 1
    elif sentence[i] != sentence[i+1]:# 현재 문자와 다음 문자가 다르면 count_sentence에 결과 저장
        count_sentence.append((count, sentence[i]))
        count = 1
    if i == len(sentence) - 2: # 초기값의 같은 문자열 길이가 1개 일 때 결과 저장
        count_sentence.append((count, sentence[i + 1]))

result = "" # 결과 출력 변수 result

for i in range(len(count_sentence)):
    result = result + str(count_sentence[i][1]) + str(count_sentence[i][0])
print(result)

a3b2c6a1


In [ ]:
# [Q14] Code-3
sentence = "aaabbcccccca" # 초기값 설정
result = sentence[0]      # 결과 저장
count = 0                 # 문자 카운팅

for i in sentence:
    if i == result[-1]:           # else에서 result 변수에 값 추가 -> 마지막 문자와 비교
        count += 1
    else:
        result += str(count) + i  # 마지막 문자와 i가 다를 경우 카운팅 값을 문자열 형태로 입력
        count = 1
result += str(count)              # result 변수에 카운팅 값을 문자열 형태로 추가

print(result)

a3b2c6a1


In [ ]:
# [Q14] Code-4
def sentence_string(s):
    sentence = ""   # 문자열 감지
    count = 0       # 문자 카운팅
    result = ""     # 결과 저장
    for st in s:
        if st!=sentence:  # 다른 문자열 => count 초기화
            sentence = st
            if count: result += str(count)
            result += st
            count = 1
        else:             # 동일한 문자열 => count 값 증가
            count +=1
    if count: result += str(count)
    return result

print (sentence_string("aaabbcccccca"))  

a3b2c6a1


## Q15. Duplicate Numbers
> 0 ~ 9의 문자로 된 숫자를 입력받았을 때, 이 입력값이 0 ~ 9의 모든 숫자를 각각 한 번씩만 사용한 것인지 확인하는 함수를 작성하시오.
```
입력 예시: 0123456789 01234 01234567890 6789012345 012322456789
출력 예시: true false false true false
```


In [ ]:
# [Q15] Code-1
def Num_inputs(s):
    list_s=s.split()
    for i in list_s :
        if list("".join(i)).count("0") == 1:
            if eval("+".join(i))==45:
                print("True")
            else :
                print("False")
        else :
            print("False")

Num_inputs("0123456789 01234 01234567890 6789012345 012322456789")


True
False
False
True
False


In [ ]:
# [Q15] Code-2
def check_number(num):
    for i in range(10):
        if num.count(str(i)) != 1:
            print("False")
            return

    print("True")

check_number('0123456789')
check_number('01234')
check_number('01234567890')
check_number('6789012345')
check_number('012322456789')


True
False
False
True
False


In [ ]:
# [Q15] Code-3
def chk_Num(s):
    result=[]
    for number in s:
        if number not in result:
            result.append(number)
        else:
            return False
    return len(result) == 10

print(chk_Num("0123456789"))    
print(chk_Num("01234"))           
print(chk_Num("01234567890"))     
print(chk_Num("6789012345"))      
print(chk_Num("012322456789"))


True
False
False
True
False


## Q16. 모스 부호 해독
> 문자열 형식으로 입력받은 모스 부호(dot:. dash:-)를 해독하여 영어 문장으로 출력하는 프로그램을 작성하시오.

- 글자와 글자 사이는 공백 1개, 단어와 단어 사이는 공백 2개로 구분한다.
- 예를 들어 다음 모스 부호는 "HE SLEEPS EARLY"로 해석해야 한다.
```
.... .  ... .-.. . . .--. ...  . .- .-. .-.. -.--
```
> 모스부호 규칙 표
```
문자 부호	문자   부호
A	 .-	    N	    -.
B	 -...	O	    ---
C	 -.-.	P	    .--.
D	 -..	Q	    --.-
E	 .	    R	    .-.
F	 ..-.	S	    ...
G	 --.	T	    -
H	 ....	U	    ..-
I	 ..	    V	    ...-
J	 .---	W	    .--
K	 -.-	X	    -..-
L	 .-..	Y	    -.--
M	 --	    Z	    --..
```

In [ ]:
# [Q16] Code-1
dic = {
    '.-':'A','-...':'B','-.-.':'C','-..':'D','.':'E','..-.':'F',
    '--.':'G','....':'H','..':'I','.---':'J','-.-':'K','.-..':'L',
    '--':'M','-.':'N','---':'O','.--.':'P','--.-':'Q','.-.':'R',
    '...':'S','-':'T','..-':'U','...-':'V','.--':'W','-..-':'X',
    '-.--':'Y','--..':'Z'
}

def morse(code):
    result = []
    for word in code.split("  "):
        for char in word.split(" "):
            result.append(dic[char])
        result.append(" ")
    return "".join(result)


print(morse('.... .  ... .-.. . . .--. ...  . .- .-. .-.. -.--'))


HE SLEEPS EARLY 


In [ ]:
# [Q16] Code-2
dic = {'':' ','.-':'A','-...':'B','-.-.':'C','-..':'D','.':'E','..-.':'F','--.':'G','....':'H','..':'I','.---':'J','-.-':'K','.-..':'L','--':'M','-.':'N','---':'O','.--.':'P','--.-':'Q','.-.':'R','...':'S','-':'T','..-':'U','...-':'V','.--':'W','-..-':'X','-.--':'Y','--..':'Z'}
morse=".... .  ... .-.. . . .--. ...  . .- .-. .-.. -.--"
word=morse.split(" ")
for i in word:
    print(dic[i],end="")

HE SLEEPS EARLY



---

# 다음 정규식 문제는 기말고사 후에 도전해 보세요.

네엡

## Regular Expression
- import re

## Q17. 기초 메타 문자
> 다음 중 정규식 a[.]{3,}b과 매치되는 문자열은 무엇일까?


```
acccb
a....b
aaab
a.cccb
```


In [ ]:
# [Q17] Code



## Q18. 문자열 검색
> 다음 코드의 결괏값은 무엇일까?

```
>>> import re
>>> p = re.compile("[a-z]+")
>>> m = p.search("5 python")
>>> m.start() + m.end()
```


In [ ]:
# [Q18] Code


## Q19. 그루핑
> 다음과 같은 문자열에서 휴대폰 번호 뒷자리인 숫자 4개를 ####로 바꾸는 프로그램을 정규식을 사용하여 작성하시오.

```
park 010-9999-9988
kim 010-9909-7789
lee 010-8789-7768
```


In [ ]:
# [Q19] Code


## Q20. 전방 탐색
> 다음은 이메일 주소를 나타내는 정규식이다. 이 정규식은 park@naver.com, kim@daum.net, lee@myhome.co.kr 등과 매치된다. 긍정형 전방 탐색 기법을 사용하여 .com, .net이 아닌 이메일 주소는 제외시키는 정규식을 작성하시오.

```
.*[@].*[.].*$
```


In [ ]:
# [Q20] Code


## 코딩 연습 결과를 github에 올리기
> 완성된 ipython 노트북 파일을 각자의 github에 올리시오.

- github 계정 만들기 (계정이 없는 학생)
    1. github에 sign up (Google gmail 계정을 이용해서 만드시오.)
- github에 노트 올리기
    2. github에 각자의 id로 repo를 만드시오. (한번 만들면 계속 재사용)
    3. github repo에 연습결과 노트를 올리시오.
        - 제출노트: ch8_exercises_id.ipynb
